torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -59,10 +60,24 @@ std::map<at::ScalarType, ncclDataType_t> ncclDataType = {
 
 // Helper function that gets the data type and issues error if not supported
 ncclDataType_t getNcclDataType(at::ScalarType type) {
-  try {
-    return ncclDataType.at(type);
-  } catch (std::out_of_range& e) {
-    throw std::runtime_error("Unsupported data type for NCCL process group");
+  auto it = ncclDataType.find(type);
+  if (it == ncclDataType.end()) {
+    // Input tensor is of an unsupported type.
+    auto err = c10::str(
+        "Input tensor data type is not supported for NCCL process group: ",
+        type);
+    TORCH_CHECK(false, err);
+  }
+  return it->second;
+}
+
+// Helper function that casts tensors from bool to long.
+void castTensorsFromBool(std::vector<at::Tensor>& tensors) {
+  for (auto& tensor : tensors) {
+    // TODO: a simple tensor = tensor.to(long) won't work here. The allreduce
+    // will be correct, but the modified tensor won't be reflected in Python.
+    auto asIntBufTensor = tensor.to(at::kLong);
```

Comment:
any reason for using `at::kLong` instead of `at::kByte` or `at::kChar`? They correspond to `ncclInt8` and `ncclUint8` respectively.

Reply:
The main reason is to prevent overflow issues, I think both sizes might be too small for very large-scale training use cases. Do we know an estimate of the largest no. of process groups we've seen?

Reply:
I see. I wonder if it makes sense to just support min (all) and max(any) for bool tensors? Sum and prod do not seem to apply here?

Reply:
I was going by Pieter's suggestion in a comment on this issue: https://github.com/pytorch/pytorch/issues/24137
In boolean logic, since sum means OR and product means AND, it might be reasonable for users to expect PyTorch to obey this? i.e., this would mean calling a sum reduction on boolean collectives means doing an OR of them. 

Open to either approach though.

Reply:
> since sum means OR and product means AND, it might be reasonable for users to expect PyTorch to obey this?

I see. This makes sense. My concern was converting bool into long might be a perf hit. How about we just use the bool tensor, and let applications deal with the overflow issue, and highlight that in the doc? They do have options to avoid overflow by using MIN/MAX. 

Another option is we internally replace sum with max and prod with min, as we expect true + true = true and true * false = false, etc.

Reply:
Sure, so it seems that this is best approach with:
1) Add a type apping of at::kBool to ncclUint8
2) I'm assuming (need to check) this means that nccl will treat the payload as a buffer of uint8. 
3) If the data type is of at::kBool, override the ncclOp if it's sum or product to be max or min, respectively (this will prevent the overflow issue)
4) With the above changes the reduction should work transparently.

Reply:
Updated the diff to use this approach.